drivers/sensor/MLX90640.zig

```diff
@@ -0,0 +1,882 @@
+//!
+//! Not all functionality from the datasheet is exposed in this driver
+//!
+//! The calculations were copied from https://github.com/netzbasteln/MLX90640-Thermocam
+//!
+const std = @import("std");
+const mdf = @import("../framework.zig");
+
+const Mlx90649Error = error{
+    BadPixels,
+};
+
+pub const MLX90640_Config = struct {
+    i2c: mdf.base.Datagram_Device,
+    clock: mdf.base.Clock_Device,
+    open_air_shift: f32 = 8,
+    emissivity: f32 = 0.95,
+};
+
+pub const MLX90640 = struct {
+    const Self = @This();
+
+    const parameters = struct {
+        kVdd: i16 = 0,
+        vdd25: i16 = 0,
+        KvPTAT: f32 = 0,
+        KtPTAT: f32 = 0,
+        vPTAT25: i16 = 0,
+        alphaPTAT: f32 = 0,
+        gainEE: i16 = 0,
+        tgc: f32 = 0,
+        cpKv: f32 = 0,
+        cpKta: f32 = 0,
+        resolutionEE: u8 = 0,
+        calibrationModeEE: u8 = 0,
+        KsTa: f32 = 0,
+        ksTo: [5]f32 = undefined,
+        ct: [5]i16 = undefined,
+        alpha: [768]u16 = undefined,
+        alphaScale: u8 = 0,
+        offset: [768]i32 = undefined,
+        kta: [768]i8 = undefined,
+        ktaScale: u8 = 0,
+        kv: [768]i8 = undefined,
+        kvScale: u8 = 0,
+        cpAlpha: [2]f32 = undefined,
+        cpOffset: [2]i16 = undefined,
+        ilChessC: [3]f32 = undefined,
+        brokenPixels: [5]u16 = undefined,
+        outlierPixels: [5]u16 = undefined,
+    };
+
+    eeprom: [832]u16 = undefined,
+    frame: [834]u16 = undefined,
+    frame_data: [834 * 2]u8 = undefined,
+    scratch_data: [768]f32 = undefined,
+    i2c: mdf.base.Datagram_Device,
+    clock_device: mdf.base.Clock_Device,
+    params: parameters,
+    emissivity: f32,
+    open_air_shift: f32,
+
+    scale_alpha: f32 = 0.000001,
+    control_register_address: u16 = 0x800D,
+    status_register_address: u16 = 0x8000,
+    ram_address: u16 = 0x0400,
+    eeprom_address: u16 = 0x2400,
+    device_id_1_address: u16 = 0x2407,
+
+    frame_loop: [2]u1 = [2]u1{ 0, 1 },
+    refresh_rate_mask: u16 = 0b111 << 7,
+
+    pub fn init(cfg: MLX90640_Config) !Self {
+        var obj = Self{
+            .i2c = cfg.i2c,
+            .clock_device = cfg.clock,
+            .open_air_shift = cfg.open_air_shift,
+            .emissivity = cfg.emissivity,
+            .params = parameters{},
+        };
+
+        // the temperatures are all wack on first read
+        try obj.loadFrame();
+
+        return obj;
+    }
+
+    fn write(self: *Self, address: u16, val: u48) !void {
+        var req = [4]u8{
+            @as(u8, @truncate(address >> 8)),
+            @as(u8, @truncate(address & 0xFF)),
+            @as(u8, @truncate(val >> 8)),
+            @as(u8, @truncate(val & 0xFF)),
+        };
+
+        try self.i2c.write(&req);
+    }
+
+    fn writeThenRead(self: *Self, address: u16, buf: []u16) !void {
+        var req = [2]u8{ @as(u8, @truncate(address >> 8)), @as(u8, @truncate(address & 0xFF)) };
```

Comment:
you can do `std.mem.nativeToBig` here, then do `std.mem.asBytes` to convert it to a byte slice

Reply:
Cool, I was hoping to learn better ways of doing some of this stuff.